### PR TITLE
[PLAY-962] Multi Level Select - Single Select Variant

### PR DESCRIPTION
**Pull request by gavhuang**
**What does this PR do?**
- [Runway](https://nitro.powerhrg.com/runway/backlog_items/PLAY-962) - This is only part one: add radio buttons instead of checkboxes
- [Alpha](https://github.com/powerhome/nitro-web/pull/35016)
- Adds a `single` variant to the Multi Level Select kit
  - Only one option should ever be selected
  - If there is a default selection via `selectedIds`, and there are with multiple `id`s in the array, the first `id` is selected as the default, and the rest ignored
- Adds a new prop, `inputName`, which is required for a radio group so that only one option can be selected 
- Address some linting errors and warnings

**How to test?**
1. Go to the Playbook docs and see the new "Single Select" variant
2. See Runway story for Nitro testing

#### Checklist:
- [X] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [X] **DEPLOY** I have added the `milano` label to show I'm ready for a review.
- [ ] **TESTS** I have added test coverage to my code.

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx [hunk 1]
```diff
@@ -232,9 +249,35 @@ const MultiLevelSelect = (props: MultiLevelSelectProps) => {
...
+    onSelect(selectedItem)
+  };
+
+  // Single select: reset the tree state upon typing
+  const handleRadioInputChange = (inputText: string) => {
+    modifyRecursive(formattedData, false)
+    setSingleSelectInputValue(inputText)
+    setFilterItem(inputText)
+  };
+
```
Comment (gavhuang) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:277]: This is the heart of the logic.

Upon a radio selection via the dropdown:
- Uncheck everything in the tree via `modifyRecursive`
- Find the selected item via `modifyValue`, and check it. None of its children should be selected because only one option can ever be selected
- Display the value via the `setSingleSelectInputValue`
- Reset the filter - per the mock, all dropdown options should be shown
- Close the dropdown
- Call `onSelect`

Upon text input change:
- Reset the tree and uncheck everything
- TLDR: don't break type ahead

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx [hunk 2]
```diff
@@ -78,6 +84,11 @@ const MultiLevelSelect = (props: MultiLevelSelectProps) => {
   useEffect(() => {
     if (returnAllSelected) {
       setReturnedArray(getCheckedItems(formattedData))
+    } else if (variant === "single" && selectedIds?.length) {
+      // The default is the first selectedId if there are more than one
+      const defaultSelection = filterFormattedDataById(formattedData, selectedIds[0])
+      setSingleSelectInputValue(defaultSelection[0]?.value)
+      setDefaultReturn(defaultSelection)
```
Comment (gavhuang) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:91]: Here we prevent multiple default options and populate the text input with the default option.

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx [hunk 3]
```diff
@@ -122,14 +133,20 @@ const MultiLevelSelect = (props: MultiLevelSelectProps) => {
       if (item.id != id) item.children = modifyValue(id, item.children, check)
       else {
         item.checked = check
-        item.children = modifyRecursive(item.children, check)
+
+        if (variant === "single") {
+          // Single select: no children should be checked
+          item.children = modifyRecursive(item.children, !check)
```
Comment (gavhuang) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:139]: Previously `modifyValue` checks the children - but we don't want that. A radio selection only has one selection.

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx [hunk 4]
```diff
@@ -313,68 +377,121 @@ const MultiLevelSelect = (props: MultiLevelSelectProps) => {
...
+      <div
+          className="wrapper"
+          ref={dropdownRef}
+      >
+        <div
+            className="input_wrapper"
+            onClick={handleInputWrapperClick}
+        >
+          <div className="input_inner_container">
+            {variant === "single" && defaultReturn.length !== 0
```
Comment (gavhuang) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:396]: This logic is for the hidden input in Rails.

File: playbook/app/pb_kits/playbook/pb_multi_level_select/docs/_multi_level_select_single.jsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ],
+  },
+];
+
+const MultiLevelSelectSingle = (props) => {
+  return (
+    <div>
+      <MultiLevelSelect
+          id="multiselect-single"
+          inputName="Power"
```
Comment (gavhuang) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/docs/_multi_level_select_single.jsx:77]: This new prop, `inputName` is needed so that in a radio group, only one is selected.

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx (hunk 4 above)
Comment (nidaqg) [reply playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:396]: looks correct but I couldn't find the hidden input when you inspect the input when a selection has been made, not sure what we're missing here but we can take a look!

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx (hunk 1 above)
Comment (nidaqg) [reply playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:277]: looks good @gavhuang ! Thank you for all the comments explaining your thought processes!

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx (hunk 4 above)
Comment (gavhuang) [reply playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:404]: Just pushed up a fix.

File: playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx [hunk 6]
```diff
@@ -78,6 +89,13 @@ const MultiLevelSelect = (props: MultiLevelSelectProps) => {
   useEffect(() => {
     if (returnAllSelected) {
       setReturnedArray(getCheckedItems(formattedData))
+    } else if (variant === "single") {
+      if (singleSelectedItem?.id?.length === 0) {
+        setSingleSelectedItem({ id: [], value: "", item: []})
+        setDefaultReturn(defaultReturn)
```
Comment (nidaqg) [inline playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:95]: not sure on this one @gavhuang , setting the state with it's own initial state doesn't seem like best practice. Why was this needed?

Comment (gavhuang) [reply playbook/app/pb_kits/playbook/pb_multi_level_select/_multi_level_select.tsx:95]: Going to refactor this!

**Comment by gavhuang** [comment]
@nidaqg going to mark this ready for code review. I am spinning up the alpha test but in the meantime we can chat about the code if you have questions.

**Comment by gavhuang** [comment]
@nidaqg [the Rails component works!](https://github.com/powerhome/nitro-web/pull/35016/commits/f4c1abe050f3472634abde1e0ab0d5e6b019b6e9) 🚀

We should change [the value in the `org_level_api.rb`](https://github.com/powerhome/nitro-web/blob/master/components/directory/lib/directory/org_level_api.rb#L127) to `node.name`. By doing so, we will match how we typically want our tree structured per our doc examples:

Screenshot 2023-10-05 at 6 16 39 PM

...and as an example in the Rails console (mismatch because it uses the `id` not `node.name`):

Screenshot 2023-10-05 at 6 17 09 PM

The last unrelated issue in `dev_docs/lists/normal` is that the Defaults button does not work. But that is true in the original Multi Level Select.

**Comment by nidaqg** [comment]
> [the value in the `org_level_api.rb`](https://github.com/powerhome/nitro-web/blob/master/components/directory/lib/directory/org_level_api.rb#L127)

Great work @gavhuang !
